atomic_cell.hh

```diff
@@ -35,6 +35,7 @@
 #include "utils/fragmented_temporary_buffer.hh"
 
 #include "serializer.hh"
+#include "cql3/values.hh"
```

Comment:
This is a no-no in terms of layering. cql3 is a higher level that uses atomic cells, not the other way round.

If cql3/values is sufficiently generic, we can move its contents to the basic mutation layer and remove it from cql3. So this by itself doesn't invalidate the approach.

Reply:
This could be mended by making atomic_cell::make_live a template (with the view type as a parameter). Though this would require either moving its definition from atomic_cell.cc to atomic_cell.hh, or instantiating the templates explicitly.

> If cql3/values is sufficiently generic, we can move its contents to the basic mutation layer and remove it from cql3.

cql3/values is not very generic, but it also doesn't depend on anything else in cql3, so maybe a *tiny* layering violation is better than adding templates? (Yeah, yeah, probably not).

> layering

What's that? Is that a C++26 feature? And why does types.cc include compaction_garbage_collector.hh again?

Reply:
> This could be mended by making atomic_cell::make_live a template (with the view type as a parameter). Though this would require either moving its definition from atomic_cell.cc to atomic_cell.hh, or instantiating the templates explicitly.

Or, use the type-erasing iterator I proposed. A template would be fine, though.

> 
> > If cql3/values is sufficiently generic, we can move its contents to the basic mutation layer and remove it from cql3.
> 
> cql3/values is not very generic, but it also doesn't depend on anything else in cql3, so maybe a _tiny_ layering violation is better than adding templates? (Yeah, yeah, probably not).
> 
> > layering
> 
> What's that? Is that a C++26 feature? 

:)

But let's not make things worse than they are.

> And why does types.cc include compaction_garbage_collector.hh again?

#8232.

Reply:
Yet another option is to make the constructor accept the underlying fragment iterator, instead of the cql3 wrapper.